### PR TITLE
SFU improvements

### LINKED ISSUE #4932: HTML5 Screenshare: media does not appear to viewers or presenters ocasionally.
It seems to be a race condition related to ICE exchange process. I`ll update this when we have more info on the problem.

What you see: the presentation disappears, as if the screenshare was successfully started, but the screenshare view stays blank.

screenshare-fail

### LINKED ISSUE #5227: Screenshare leaves lingering ffmpeg processes
In some cases there are ffmpeg processes on the server that are left over from screenshare that was long gone

### PR DESCRIPTION
**Pull request by prlanzarin**
This is another batch of improvements to the SFU app, including:
- Added ICE queues to the Screenshare manager. Should fix #4932, but gotta test this under high load to be sure.
- Graceful exiting processes. This should avoid lingering ffmpeg processes and inconsistent meeting state on screensharing (fixes #5227 and #5219). And also clean up things properly on kurento when we shut down bbb-webrtc-sfu (improves the situation on #4990, and I think that's the most we can do regarding this for now).
  - To check for Kurento's stale endpoints/pipelines, you can use this script https://gist.github.com/fcecagno/8ed9e9ae2bb7aded329f01a2acbeff60. Just run `npm install` and wrap it under a bash script to keep it running (e.g.: ` while true; do npm start; sleep 1; done;`)
- Fixed video viewer session stop. The client wasn't sending the `stop` message to the SFU app.
- Made the H264 enforcement configurable on SFU. Look for `webcam-force-h264` and `screenshare-force-h264` in `default.example.yml`.
- Detect kurento disconnection and propagate that to clients. The behaviour right now is to stop everything and leave things in a consistent state. Reconnections (for video/screenshare) when Kurento goes back up in come in another PR.
- Auto reconnection to Kurento (we can bring kurento down and up without having to restart SFU)
- Some more work on error handling; we should see fewer unhandled errors, but it's still a WIP. I'll be focusing on a definitive refactor for that next.
- Major refactor of the video/screenshare manager and process classes. The provider classes will prolly come in another PR (I'm talking about video.js and screenshare.js). The idea with this is to make the most code I can reusable so we can fix things only once and on the same place
- Some inner refactor on the mcs-core lib

### CODE REVIEW

File: labs/bbb-webrtc-sfu/lib/mcs-core/lib/media/media-server.js [hunk 1]
```diff
@@ -177,8 +179,7 @@ module.exports = class MediaServer extends EventEmitter {
     let mediaElement = this._mediaElements[elementId];
     let kurentoCandidate = mediaServerClient.getComplexType('IceCandidate')(candidate);
 
-    if (typeof mediaElement !== 'undefined' && typeof mediaElement.addIceCandidate === 'function' &&
-        typeof candidate !== 'undefined') {
+    if (mediaElement != null  && candidate != null) {
```
Comment (lfzawacki) [inline labs/bbb-webrtc-sfu/lib/mcs-core/lib/media/media-server.js]: Don't you think it's better just to do !mediaElement in all these next cases just like you did in line 28 ?

Comment (prlanzarin) [reply labs/bbb-webrtc-sfu/lib/mcs-core/lib/media/media-server.js]: Fixed it.

File: labs/bbb-webrtc-sfu/lib/video/VideoManager.js [hunk 2]
```diff
@@ -11,201 +11,203 @@ const Video = require('./video');
...
 
-let stopAll = function() {
-  Logger.info('[VideoManager] Stopping everything! ');
+  stopAll () {
+    Logger.info('[VideoManager] Stopping everything! ');
 
-  if (sessions == null) {
-    return;
-  }
+    if (this.sessions == null) {
```
Comment (lfzawacki) [inline labs/bbb-webrtc-sfu/lib/video/VideoManager.js]: !this.sessions

File: labs/bbb-webrtc-sfu/lib/mcs-core/lib/model/SfuUser.js [hunk 3]
```diff
@@ -32,17 +33,17 @@ module.exports = class SfuUser extends User {
     // TODO switch from type to children UriSessions (RTSP|HTTP|etc)
     let session = new UriSession(uri, type);
 
-    if (typeof this._mediaSessions[session.id] == 'undefined' || 
+    if (typeof this._mediaSessions[session.id] == 'undefined' ||
```
Comment (lfzawacki) [inline labs/bbb-webrtc-sfu/lib/mcs-core/lib/model/SfuUser.js]: ooops

**Pull request merged**